crates/holochain_state/src/integrate.rs

```diff
@@ -0,0 +1,111 @@
+use holo_hash::{AnyDhtHash, DhtOpHash, HasHash};
+use holochain_p2p::HolochainP2pDnaT;
+use holochain_sqlite::rusqlite::Transaction;
+use holochain_types::{
+    dht_op::{DhtOp, DhtOpHashed, DhtOpType},
+    prelude::DhtOpResult,
+};
+use holochain_zome_types::{EntryVisibility, SignedHeader};
+
+use crate::{prelude::*, query::get_public_op_from_db};
+
+/// Insert any authored ops that have been locally validated
+/// into the dht database awaiting integration.
+/// This checks if ops are within the storage arc
+/// of any local agents.
+pub async fn authored_ops_to_dht_db(
+    network: &(dyn HolochainP2pDnaT + Send + Sync),
+    hashes: Vec<(DhtOpHash, AnyDhtHash)>,
+    authored_env: &DbRead<DbKindAuthored>,
+    dht_env: &DbWrite<DbKindDht>,
+) -> StateMutationResult<()> {
+    // Check if any agents in this space are an authority for these hashes.
+    let mut should_hold_hashes = Vec::with_capacity(hashes.len());
+    for (op_hash, basis) in hashes {
+        if network.authority_for_hash(basis).await? {
+            should_hold_hashes.push(op_hash);
```

Comment:
```suggestion
    let mut should_hold_hashes = Vec::new();
    for (op_hash, basis) in hashes {
        if network.authority_for_hash(basis).await? {
            should_hold_hashes.push(op_hash);
```

The default memory allocation for vecs is actually pretty efficient, unless you know the exact count you want, sometimes over-allocating can be slower because it's harder to find large contiguous blocks of memory.